### PR TITLE
More ergonomic spatial audio

### PR DESCRIPTION
**Pull request by rparrett**
# Objective

Spatial audio was heroically thrown together at the last minute for Bevy 0.10, but right now it's a bit of a pain to use -- users need to manually update audio sinks with the position of the listener / emitter.

Hopefully the migration guide entry speaks for itself.

## Solution

Add a new `SpatialListener` component and automatically update sinks with the position of the listener and and emitter.

## Changelog

`SpatialAudioSink`s are now automatically updated with positions of emitters and listeners.

## Migration Guide

Spatial audio now automatically uses the transform of the `AudioBundle` and of an entity with a `SpatialListener` component.

If you were manually scaling emitter/listener positions, you can use the `spatial_scale` field of `AudioPlugin` instead.

```rust

// Old

commands.spawn(
    SpatialAudioBundle {
        source: asset_server.load("sounds/Windless Slopes.ogg"),
        settings: PlaybackSettings::LOOP,
        spatial: SpatialSettings::new(listener_position, gap, emitter_position),
    },
);

fn update(
    emitter_query: Query<(&Transform, &SpatialAudioSink)>,
    listener_query: Query<&Transform, With<Listener>>,
) {
    let listener = listener_query.single();

    for (transform, sink) in &emitter_query {
        sink.set_emitter_position(transform.translation);
        sink.set_listener_position(*listener, gap);
    }
}

// New

commands.spawn((
    SpatialBundle::from_transform(Transform::from_translation(emitter_position)),
    AudioBundle {
        source: asset_server.load("sounds/Windless Slopes.ogg"),
        settings: PlaybackSettings::LOOP.with_spatial(true),
    },
));

commands.spawn((
    SpatialBundle::from_transform(Transform::from_translation(listener_position)),
    SpatialListener::new(gap),
));
```

## Discussion

I removed `SpatialAudioBundle` because the `SpatialSettings` component was made mostly redundant, and without that it was identical to `AudioBundle`.

`SpatialListener` is a bare component and not a bundle which is feeling like a maybe a strange choice. That happened from a natural aversion both to nested bundles and to duplicating `Transform` etc in bundles and from figuring that it is likely to just be tacked on to some other bundle (player, head, camera) most of the time.

Let me know what you think about these things / everything else.

### CODE REVIEW

File: crates/bevy_audio/src/audio.rs
Comment (nerdachse) [inline crates/bevy_audio/src/audio.rs:183]: Great, there goes my awesome idea of an alien species game that have ears on their front and back sides! Pah! :)

Comment (rparrett) [reply crates/bevy_audio/src/audio.rs:183]: That's no problem -- you can still construct `SpatialListener` manually with your own ear offsets. This constructor is just for convenience with the most common use case.

**Review by hymm — APPROVED** [review]
Made some comments, but nothing blocking. This API definitely makes more sense to me than the status quo.

File: crates/bevy_audio/src/audio_output.rs
Comment (rparrett) [reply crates/bevy_audio/src/audio_output.rs:286]: Actually I think `update_listener_positions` looping over changed emitters is a mistake.

This should update all emitters when the listener has changed.

Comment (rparrett) [reply crates/bevy_audio/src/audio_output.rs:289]: I updated that system so that it's only doing work when listeners change.

**Comment by rparrett** [comment]
I fixed a few major issues that weren't showing up in the examples and expanded the examples a bit so that you can move the listener around.

This has changed enough that previous approvals should probably be invalidated.

File: crates/bevy_audio/src/audio_output.rs
Comment (mockersf) [inline crates/bevy_audio/src/audio_output.rs:65]: this should use `get_single`

Comment (mockersf) [reply crates/bevy_audio/src/audio_output.rs:65]: warning is already handled somewhere else 👍

File: crates/bevy_audio/src/audio_output.rs
Comment (mockersf) [inline crates/bevy_audio/src/audio_output.rs:123]: "first" is not very helpful. Could you log the entity used?

File: crates/bevy_audio/src/audio_output.rs
Comment (rparrett) [reply crates/bevy_audio/src/audio_output.rs]: For anyone else reading, there was some discussion about this over on discord: https://discord.com/channels/691052431525675048/1158058216882712677

I added some comment. Let me know what you think.

File: crates/bevy_audio/src/audio_output.rs
Comment (rparrett) [reply crates/bevy_audio/src/audio_output.rs]: Good idea!

**Comment by rparrett** [comment]
Sorry for the noise. I'm done fiddling with this now.

File: crates/bevy_audio/src/audio.rs
Comment (hymm) [inline crates/bevy_audio/src/audio.rs:164]: should delete this empty line and the one above it.

File: crates/bevy_audio/src/audio_output.rs
Comment (hymm) [inline crates/bevy_audio/src/audio_output.rs:64]:
```suggestion
    /// If there are no listeners, use the default values. If a user has added multiple
```

File: crates/bevy_audio/src/audio_output.rs
Comment (hymm) [inline crates/bevy_audio/src/audio_output.rs:66]:
```suggestion
    /// listeners for whatever reason, we will return the first value.
```

**Review by hymm — APPROVED** [review]
Approving, but my nits should probably be fixed before merging.

File: crates/bevy_audio/src/audio.rs
Comment (rparrett) [reply crates/bevy_audio/src/audio.rs]: Those docs on this and `new` were a bit funky with leftover bits from the old implementation. Should be making more sense now.

**Review by mockersf — APPROVED** [review]

**Pull request merged**
